Push partial aggregation through inner join

pnowojski opened the pull request:
Adding flag for pushing partial aggregation through join.

This is yet another PR that depends on jmh benchmarks defined in https://github.com/prestodb/presto/pull/5809. Results for simple query:
```
Benchmark                                                       (enabled)  Mode  Cnt    Score    Error  Units
BenchmarkPartialAggregationPushdown.pushdownPartialThroughJoin       true  avgt   10  210.992 ±  7.088  ms/op
BenchmarkPartialAggregationPushdown.pushdownPartialThroughJoin      false  avgt   10  494.724 ± 58.796  ms/op
```

With support for `filter function` (there is now `TODO` in the code for this) this change would significantly outperform SPARK 2.0 in [CERN use case](https://databricks.com/blog/2016/10/03/voice-from-cern-apache-spark-2-0-performance-improvements-investigated-with-flame-graphs.html)

pnowojski commented:
@kokosing @martint I have tried to rewrite this rule to iterative optimizer but didn't see how can I use session variables there. I hope that this is the missing feature?

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PartialAggregationPushDown.java:
```diff
@@ -140,6 +158,147 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
+                    aggregation.getStep(),
+                    aggregation.getHashSymbol(),
+                    aggregation.getGroupIdSymbol());
+            return new SymbolMapper(mapping.build()).map(pushedAggregation, source);
```
To make sure I understand:  The idea behind this mapping here is that anything in the grouping set that referenced  e.g. the left side of the join clause if we're pushing right gets rewritten to be about the right side. And we can do that because they have an equality predicate between them.

pnowojski replied:
Yes, that's correct.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java:
```diff
@@ -0,0 +1,131 @@
+
+    public Symbol map(Symbol symbol)
+    {
+        Symbol canonical = symbol;
```
Don't call the mapped symbol "canonical"  since you use it more generically for any mapping between symbols.  (throughout this file)

pnowojski replied:
How do you think it should be named? `mappedSymbol`? I didn't want to rename variables, since this is just copy/paste code from `UnaliasSymbolReferences`

rschlussel replied:
mappedSymbol is fine

pnowojski commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -129,35 +129,9 @@ private Rewriter(Map<Symbol, Type> types)
-                    node.getStep(),
-                    canonicalize(node.getHashSymbol()),
-                    canonicalize(node.getGroupIdSymbol()));
+            //TODO: use mapper in other methods
```
I could, but this is additional work, that we don't know whether will be required at all.

pnowojski commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -52,6 +52,7 @@
     public static final String QUERY_MAX_CPU_TIME = "query_max_cpu_time";
     public static final String REDISTRIBUTE_WRITES = "redistribute_writes";
     public static final String PUSH_TABLE_WRITE_THROUGH_UNION = "push_table_write_through_union";
+    public static final String PUSH_AGGREGATION_THROUGH_JOIN = "push_aggregation_through_join";
```
Maybe we should name it `push_partial_aggregation_through_join` here?

pnowojski commented on presto-tests/src/test/java/com/facebook/presto/tests/TestLocalQueries.java:
```diff
@@ -39,6 +40,7 @@ public static LocalQueryRunner createLocalQueryRunner()
                 .setCatalog("local")
                 .setSchema(TINY_SCHEMA_NAME)
                 .setSystemProperty(REORDER_JOINS, "true")
+                .setSystemProperty(PUSH_AGGREGATION_THROUGH_JOIN, "true")
```
I don't know. IMO in this case, enabling here has more advantages then leaving it disabled. Turning this on will likely to test all test cases, while disabling it will never test partial aggregations under joins.

Ideally, we should test both versions, but this is a topic for another story: how do we want to test features? Testing all combination is impossible, but maybe we want `TestLocalQueries` with all optional features disabled and another one "feature reach" `TestLocalQueries` with all optional features enabled?

cc @martint

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PartialAggregationPushDown.java:
```diff
@@ -140,6 +158,147 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
+        {
+            Set<String> outputNames = outputSymbols.stream().map(Symbol::getName).collect(toImmutableSet());
+
+            return aggregations.values().stream()
```
This whole method can probably be replaced with something like:

```java
Set<Symbol> inputs = DependencyExtractor.extracUnique(aggregations.values());
return outputSymbols.containsAll(inputs);
```

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PartialAggregationPushDown.java:
```diff
@@ -140,6 +158,147 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
+            // keep symbols that are either directly from the join's child (availableSymbols) or there is
+            // an equality in join condition to a symbol for the join child
+            List<Symbol> pushedDownGroupingSet = groupingSet.stream()
+                    .filter((symbol) -> joinKeys.contains(symbol) || availableSymbols.contains(symbol))
```
Parenthesis not needed around the lambda argument.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PartialAggregationPushDown.java:
```diff
@@ -140,6 +158,147 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
+            List<Symbol> groupingSet = Iterables.getOnlyElement(aggregation.getGroupingSets());
+            Set<Symbol> joinKeys = Stream.concat(
+                    join.getCriteria().stream().map(EquiJoinClause::getLeft),
+                    join.getCriteria().stream().map(EquiJoinClause::getRight)).collect(Collectors.toSet());
```
Place the `.collect` on a separate line.

pnowojski commented:
I have rebased this code and renamed property. 

@rschlussel regarding multiple `INTERMEDIATE` aggregations, presto supports that, it would be sometimes beneficial, but I would add support for it once we have `CostCalculator` that works on aggregations. Otherwise we would need even more session properties to control this feature.

pnowojski commented:
Thanks :)
